addons/mail/models/mail_message.py

```diff
@@ -11,6 +11,7 @@
 
 
 _logger = logging.getLogger(__name__)
+_image_dataurl = re.compile(r'(data:image/(.+));base64,([^"\']*)', re.I)
```

Comment:
Good point about the backtracking of the first part, and the restricted character set.
`[a-z]+?` looks quite sufficient to me indeed. Plus we don't want to capture it, we're going to let the mimetype detection do its job while creating the attachment (and securely).
The base64 part could be slightly improved as `([a-z0-9+/=]{3,})` too.

According to regex101, the resulting regex `(data:image/[a-z]+?);base64,([a-z0-9+/]{3,}=*)` is ~30 times faster:
![re_orig](https://user-images.githubusercontent.com/2039870/27872143-1200263e-61a8-11e7-9b47-f20143e8fb31.png)
![re_improved](https://user-images.githubusercontent.com/2039870/27872148-167590a0-61a8-11e7-940e-1cdb019066e5.png)